src/nvme/fabrics.h

```diff
@@ -197,13 +197,52 @@ int nvmf_add_ctrl(nvme_host_t h, nvme_ctrl_t c,
  * nvmf_get_discovery_log() - Return the discovery log page
  * @c:			Discover controller to use
  * @logp:		Pointer to the log page to be returned
- * @max_retries:	maximum number of log page entries to be returned
+ * @max_retries:	Number of retries in case of failure
+ *
+ * The memory allocated for the log page and returned in @logp
+ * must be freed by the caller.
  *
  * Return: 0 on success; on failure -1 is returned and errno is set
  */
 int nvmf_get_discovery_log(nvme_ctrl_t c, struct nvmf_discovery_log **logp,
 			   int max_retries);
 
+/**
+ * nvmf_get_discovery_log2() - Return the discovery log page
+ * @c:			Discover controller to use
+ * @args:		Argument structure
+ * @logp:		Pointer to the log page to be returned
+ * @max_retries:	Number of retries in case of failure
+ *
+ * The @args parameter is typically set as follows:
+ *
+ * struct nvme_get_log_args args = {
+ * 	.args_size = sizeof(args),
+ * 	.fd = 0,			< Set internally
+ * 	.nsid = NVME_NSID_NONE,
+ * 	.lsp = NVMF_LOG_DISC_LSP_NONE,	< See enum nvmf_get_log_discovery_lsp
+ * 	.lsi = NVME_LOG_LSI_NONE,
+ * 	.uuidx = NVME_UUID_NONE,
+ * 	.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+ * 	.result = NULL,
+ * 	.lid = NVME_LOG_LID_DISCOVER,
+ * 	.log = NULL,			< Set internally
```

Comment:
Obviously if we drop the log argument, this one needs to be set correctly.

Reply:
Actually, the function dynamically allocates memory based on the responses it gets from the Discovery Controller (DC). That is, it performs a Get Discovery Log Page (DLP) using just the "minimum memory" required to Get the DLP header which contains the number of DLP Entries (DLPE) that are available at the DC. The number of DLPEs is used to determine how much memory will be required to perform a second Get DLP command. But before allocating a second larger buffer, `nvmf_get_discovery_log()` frees the "minimum memory" buffer previously allocated. The second larger buffer is the one that gets returned in `.log` and `logp`. 

Note that if the DLP changes while being retrieved (indicated by a different `genctr` between the first and second Get DLP), then the algorithm starts all over again freeing and allocating new memory. So, it's not possible for the caller to pre-allocate memory and pass it in using "`.log`".

Maybe passing `struct nvme_get_log_args args` is not the best approach for the Get DLP command. When we look at the structure (below) we see that most of the parameters are not configurable and must be set to specific values. In fact, I believe that only the `lsp`, `timeout`, and possibly the `csi` could be configured by a user when performing a Get DLP command.

```
struct nvme_get_log_args args = {
	.args_size = sizeof(args),
	.fd = 0,			< Set internally
	.nsid = NVME_NSID_NONE,
	.lsp = NVMF_LOG_DISC_LSP_NONE,	< See enum nvmf_get_log_discovery_lsp
	.lsi = NVME_LOG_LSI_NONE,
	.uuidx = NVME_UUID_NONE,
	.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
	.result = NULL,
	.lid = NVME_LOG_LID_DISCOVER,
	.log = NULL,			< Set internally
	.len = 0,			< Set internally
	.csi = NVME_CSI_NVM,
	.rae = false,			< Set internally
	.ot = false,
};
```

Let me know if I should take a different approach than passing `struct nvme_get_log_args args`. I could, for example, define a new struct containing a smaller number of arguments (only those that are truly configurable).

Reply:
The args approach seems indeed a bit wrong in this case. Alright, let's introduce a function with just the args you need. Maybe we find also a matching name along side then. It sounds like the new function is used in the scenario where have this kind of 'nested' lookups.